File: ts-portal-db/app/routers/notices.py
"""
공지사항 API 라우터
"""

from typing import List
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from ..database import get_db
from ..services.notice_service import NoticeService
from ..schemas.notice import (
    NoticeCreate,
    NoticeUpdate,
    NoticeResponse,
    NoticeListResponse,
    NoticeStats,
    SearchParams,
    PaginationParams,
    NoticePriorityEnum
)

router = APIRouter(prefix="/notices", tags=["notices"])

def get_notice_service(db: Session = Depends(get_db)) -> NoticeService:
    """공지사항 서비스 의존성"""
    return NoticeService(db)

@router.post("/", response_model=NoticeResponse, summary="공지사항 생성")
async def create_notice(
    notice_data: NoticeCreate,
    service: NoticeService = Depends(get_notice_service)
):
    """새로운 공지사항을 생성합니다."""
    try:
        return service.create_notice(notice_data)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"공지사항 생성 중 오류가 발생했습니다: {str(e)}")

@router.get("/", response_model=NoticeListResponse, summary="공지사항 목록 조회")
async def get_notices(
    skip: int = Query(0, ge=0, description="건너뛸 개수"),
    limit: int = Query(20, ge=1, le=100, description="가져올 개수"),
    service: NoticeService = Depends(get_notice_service)
):
    """공지사항 목록을 조회합니다. 고정 공지사항이 먼저 표시됩니다."""
    pagination = PaginationParams(skip=skip, limit=limit)
    return service.get_notices(pagination)

@router.get("/search", response_model=NoticeListResponse, summary="공지사항 검색")
async def search_notices(
    q: str = Query(None, description="검색 키워드"),
    priority: NoticePriorityEnum = Query(None, description="중요도 필터"),
    author_id: int = Query(None, description="작성자 ID 필터"),
    is_pinned: bool = Query(None, description="고정 공지사항만"),
    active_only: bool = Query(True, description="활성 공지사항만"),
    skip: int = Query(0, ge=0, description="건너뛸 개수"),
    limit: int = Query(20, ge=1, le=100, description="가져올 개수"),
    service: NoticeService = Depends(get_notice_service)
):
    """공지사항을 검색합니다."""
    search_params = SearchParams(
        q=q,
        priority=priority,
        author_id=author_id,
        is_pinned=is_pinned,
        active_only=active_only
    )
    pagination = PaginationParams(skip=skip, limit=limit)
    return service.search_notices(search_params, pagination)

@router.get("/pinned", response_model=List[NoticeResponse], summary="고정 공지사항 조회")
async def get_pinned_notices(
    service: NoticeService = Depends(get_notice_service)
):
    """고정된 공지사항들을 조회합니다."""
    return service.get_pinned_notices()

@router.get("/recent", response_model=List[NoticeResponse], summary="최근 공지사항 조회")
async def get_recent_notices(
    days: int = Query(7, ge=1, le=30, description="최근 며칠간의 공지사항"),
    limit: int = Query(5, ge=1, le=20, description="최대 개수"),
    service: NoticeService = Depends(get_notice_service)
):
    """최근 공지사항들을 조회합니다."""
    return service.get_recent_notices(days=days, limit=limit)

@router.get("/stats/summary", response_model=NoticeStats, summary="공지사항 통계")
async def get_notice_stats(
    service: NoticeService = Depends(get_notice_service)
):
    """공지사항 통계 정보를 조회합니다."""
    return service.get_notice_stats()

@router.get("/author/{author_id}", response_model=NoticeListResponse, summary="작성자별 공지사항 조회")
async def get_notices_by_author(
    author_id: int,
    skip: int = Query(0, ge=0, description="건너뛸 개수"),
    limit: int = Query(20, ge=1, le=100, description="가져올 개수"),
    service: NoticeService = Depends(get_notice_service)
):
    """특정 작성자의 공지사항들을 조회합니다."""
    pagination = PaginationParams(skip=skip, limit=limit)
    return service.get_notices_by_author(author_id, pagination)

@router.get("/{notice_id}", response_model=NoticeResponse, summary="공지사항 단일 조회")
async def get_notice(
    notice_id: int,
    service: NoticeService = Depends(get_notice_service)
):
    """특정 공지사항을 조회합니다."""
    notice = service.get_notice(notice_id)
    if not notice:
        raise HTTPException(status_code=404, detail="공지사항을 찾을 수 없습니다.")
    return notice

@router.put("/{notice_id}", response_model=NoticeResponse, summary="공지사항 수정")
async def update_notice(
    notice_id: int,
    notice_data: NoticeUpdate,
    service: NoticeService = Depends(get_notice_service)
):
    """공지사항을 수정합니다."""
    try:
        notice = service.update_notice(notice_id, notice_data)
        if not notice:
            raise HTTPException(status_code=404, detail="공지사항을 찾을 수 없습니다.")
        return notice
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"공지사항 수정 중 오류가 발생했습니다: {str(e)}")

@router.delete("/{notice_id}", summary="공지사항 삭제")
async def delete_notice(
    notice_id: int,
    service: NoticeService = Depends(get_notice_service)
):
    """공지사항을 삭제합니다 (소프트 삭제)."""
    try:
        success = service.delete_notice(notice_id)
        if not success:
            raise HTTPException(status_code=404, detail="공지사항을 찾을 수 없습니다.")
        return {"message": "공지사항이 성공적으로 삭제되었습니다."}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"공지사항 삭제 중 오류가 발생했습니다: {str(e)}")

@router.get("/priorities/list", summary="중요도 목록 조회")
async def get_priority_list():
    """공지사항 중요도 목록을 조회합니다."""
    return [
        {"value": "normal", "label": "일반", "icon": "📢", "color": "#6B7280"},
        {"value": "caution", "label": "주의", "icon": "⚠️", "color": "#F59E0B"},
        {"value": "important", "label": "중요", "icon": "🚨", "color": "#EF4444"}
    ] 

File: frontend-new/src/services/notices.ts
/**
 * Notice Service
 * Kong Gateway를 통한 notice-service 연동
 */

import { kongApi } from './api'
import type { 
  NoticeCreate,
  NoticeUpdate,
  NoticeResponse,
  NoticeListResponse,
  NoticeStats,
  SearchParams,
  PaginationParams,
  PriorityInfo
} from '@/types/notices'

export class NoticesService {
  private readonly baseUrl = '/api/notices'

  /**
   * 공지사항 목록 조회
   */
  async getNotices(skip: number = 0, limit: number = 20): Promise<NoticeListResponse> {
    console.log('📋 공지사항 목록 조회 요청:', { skip, limit })
    const params = new URLSearchParams({ 
      skip: skip.toString(), 
      limit: limit.toString() 
    })
    const response = await kongApi.get<NoticeListResponse>(`${this.baseUrl}/?${params}`)
    console.log('📋 공지사항 목록 조회 성공:', response.total, '개 공지사항')
    return response
  }

  /**
   * 공지사항 검색
   */
  async searchNotices(
    searchParams: SearchParams, 
    pagination: PaginationParams
  ): Promise<NoticeListResponse> {
    console.log('📋 공지사항 검색 요청:', { searchParams, pagination })
    
    const params = new URLSearchParams({
      skip: pagination.skip.toString(),
      limit: pagination.limit.toString()
    })
    
    // 검색 파라미터 추가
    if (searchParams.q) params.append('q', searchParams.q)
    if (searchParams.priority) params.append('priority', searchParams.priority)
    if (searchParams.author_id) params.append('author_id', searchParams.author_id.toString())
    if (searchParams.is_pinned !== undefined) params.append('is_pinned', searchParams.is_pinned.toString())
    if (searchParams.active_only !== undefined) params.append('active_only', searchParams.active_only.toString())
    
    const response = await kongApi.get<NoticeListResponse>(`${this.baseUrl}/search?${params}`)
    console.log('📋 공지사항 검색 성공:', response.total, '개 결과')
    return response
  }

  /**
   * 고정 공지사항 조회
   */
  async getPinnedNotices(): Promise<NoticeResponse[]> {
    console.log('📌 고정 공지사항 조회 요청')
    const response = await kongApi.get<NoticeResponse[]>(`${this.baseUrl}/pinned`)
    console.log('📌 고정 공지사항 조회 성공:', response.length, '개')
    return response
  }

  /**
   * 최근 공지사항 조회
   */
  async getRecentNotices(days: number = 7, limit: number = 5): Promise<NoticeResponse[]> {
    console.log('📅 최근 공지사항 조회 요청:', { days, limit })
    const params = new URLSearchParams({ 
      days: days.toString(), 
      limit: limit.toString() 
    })
    const response = await kongApi.get<NoticeResponse[]>(`${this.baseUrl}/recent?${params}`)
    console.log('📅 최근 공지사항 조회 성공:', response.length, '개')
    return response
  }

  /**
   * 공지사항 통계 조회
   */
  async getNoticeStats(): Promise<NoticeStats> {
    console.log('📊 공지사항 통계 조회 요청')
    const response = await kongApi.get<NoticeStats>(`${this.baseUrl}/stats`)
    console.log('📊 공지사항 통계 조회 성공')
    return response
  }

  /**
   * 작성자별 공지사항 조회
   */
  async getNoticesByAuthor(
    authorId: number, 
    skip: number = 0, 
    limit: number = 20
  ): Promise<NoticeListResponse> {
    console.log('👤 작성자별 공지사항 조회 요청:', { authorId, skip, limit })
    const params = new URLSearchParams({ 
      skip: skip.toString(), 
      limit: limit.toString() 
    })
    const response = await kongApi.get<NoticeListResponse>(`${this.baseUrl}/author/${authorId}?${params}`)
    console.log('👤 작성자별 공지사항 조회 성공:', response.total, '개')
    return response
  }

  /**
   * 특정 공지사항 조회
   */
  async getNotice(noticeId: number): Promise<NoticeResponse> {
    console.log('📋 공지사항 조회 요청:', noticeId)
    const response = await kongApi.get<NoticeResponse>(`${this.baseUrl}/${noticeId}`)
    console.log('📋 공지사항 조회 성공:', response.title)
    return response
  }

  /**
   * 공지사항 생성
   */
  async createNotice(noticeData: NoticeCreate): Promise<NoticeResponse> {
    console.log('📝 공지사항 생성 요청:', noticeData.title)
    const response = await kongApi.post<NoticeResponse>(`${this.baseUrl}/`, noticeData)
    console.log('📝 공지사항 생성 성공:', response.id)
    return response
  }

  /**
   * 공지사항 수정
   */
  async updateNotice(noticeId: number, noticeData: NoticeUpdate): Promise<NoticeResponse> {
    console.log('✏️ 공지사항 수정 요청:', noticeId)
    const response = await kongApi.put<NoticeResponse>(`${this.baseUrl}/${noticeId}`, noticeData)
    console.log('✏️ 공지사항 수정 성공:', response.id)
    return response
  }

  /**
   * 공지사항 삭제
   */
  async deleteNotice(noticeId: number): Promise<{ message: string }> {
    console.log('🗑️ 공지사항 삭제 요청:', noticeId)
    const response = await kongApi.delete<{ message: string }>(`${this.baseUrl}/${noticeId}`)
    console.log('🗑️ 공지사항 삭제 성공:', noticeId)
    return response
  }

  /**
   * 우선순위 목록 조회
   */
  async getPriorities(): Promise<PriorityInfo[]> {
    console.log('📊 우선순위 목록 조회 요청')
    const response = await kongApi.get<PriorityInfo[]>(`${this.baseUrl}/priorities/list`)
    console.log('📊 우선순위 목록 조회 성공:', response.length, '개')
    return response
  }

  /**
   * 날짜 포맷팅 유틸리티
   */
  formatDate(dateString: string): string {
    if (!dateString) return '날짜 정보 없음'
    
    try {
      const date = new Date(dateString)
      const now = new Date()
      const diffInHours = Math.floor((now.getTime() - date.getTime()) / (1000 * 60 * 60))
      
      if (diffInHours < 1) return '방금 전'
      if (diffInHours < 24) return `${diffInHours}시간 전`
      if (diffInHours < 24 * 7) return `${Math.floor(diffInHours / 24)}일 전`
      
      return date.toLocaleDateString('ko-KR', {
        year: 'numeric',
        month: 'short',
        day: 'numeric',
        hour: '2-digit',
        minute: '2-digit'
      })
    } catch (error) {
      return '날짜 정보 없음'
    }
  }

  /**
   * 우선순위 색상 가져오기
   */
  getPriorityColor(priority: string): string {
    const colors = {
      normal: '#6B7280',
      caution: '#F59E0B',
      important: '#EF4444'
    }
    return colors[priority as keyof typeof colors] || '#6B7280'
  }

  /**
   * 우선순위 아이콘 가져오기
   */
  getPriorityIcon(priority: string): string {
    const icons = {
      normal: '📢',
      caution: '⚠️',
      important: '🚨'
    }
    return icons[priority as keyof typeof icons] || '📢'
  }
}

// 싱글톤 인스턴스 생성
export const noticesService = new NoticesService()

// 기본 export
export default noticesService 